### ISSUE TITLE
Feature Request: Limit Host Scheduling

### ISSUE THREAD

**Issue by joshbenner** [rancher/rancher#7169]
Allow one to specify scheduling filters on the host. For instance, "only schedule containers with label foo=bar".

This could help several use cases:
- [Exclude hosts from scheduling - #7168](7168) (effectively?)
- A single env being across multiple DCs, but wanting affinities of certain services to certain DCs
- Keeping services in proper legal jurisdictions
- Dedicated hosts in environment (ie: "the DB servers" or whatnot)

**Comment by vincent99** [comment]
As I said in IRC I think separate environments are the correct mechanism to do this.  Besides providing better isolation (different users can be given access, separate overlay network, etc) it already exists and gives you a very explicit separation.  Using labels to exclude or include from both directions will gives you separation you have to squint to find.  It will be very difficult for the user to tell what's going on and very easy to end up in situations where no hosts are eligible because you added *a* label the special host has but not *the right* one to let it be scheduled there.

To do multiple environments we basically just need to add functionality to make that setup usable... namely a way to link to a service across environment boundaries And have the container join the other overlay network.

**Comment by ekristen** [comment]
It seems to me that this is the wrong approach. Would it not be better to introduce some other mechanism for better user controls? 

An environment is designed to logical group servers together so that you can link services together and load balance services, etc ...

In this context, would it not be best to keep all servers whether it be cross cloud provider or in the same provider in the same Rancher environment?

For example, if I have 3 groupings of servers, application, management, services, but I want and need services across the 3 groupings to talk to each other, but control who can access what, I can't do this unless I follow your recommendation. :(

ITSM by using your recommendation to use environments for permission controls is ANTI Rancher.

**Comment by VAdamec** [comment]
Agree with @ekristen affinity labels on hosts level make sense in a lot of situation, for example you need to isolate where balancers (haproxy/traefic/..) need to run. Best approach is to have hosts just for this purpose and you don't want o to run anything else here. Right now you need to set affinity rule on all services to avoid this. Rule on hosts will be definitely better and more secure. How separate environment helps me with this ? If I place balancers to different environment I loose access to events from environment where services running, how balancer can work correctly then ?

**Comment by beatcracker** [comment]
+1 on this. I need to exclude a single host and this means updating **all** my compose files with scheduling labels. Not nice.

**Comment by vincent99** [comment]
We already added this in 1.5 (and as I said it's caused a fair amount of confusion when used).  "Require container label" in edit host.

**Comment by ekristen** [comment]
This is a great feature glad to see it in rancher!

I would argue the confusion is that Rancher does a very poor job of indicating why a service is not getting scheduled.

I recently ran into this while *NOT* using this feature.  I was in fact using the scheduling feature already built for a service to require a host with a specific label, and I did not have any hosts with that label. Therefore the services just sat there showing unhealthy and no containers. The service logs basically just said "everything was reconciled".
